alf/algorithms/ppg/disjoint_policy_value_network.py

```diff
@@ -0,0 +1,244 @@
+
+        # Note that the aux branch value head belongs to the policy component.
+
+        # Like the value head Aux head is outputing value estimation
+        self._aux_head = alf.nn.Sequential(
+            alf.layers.FC(
+                input_size=self._actor_encoder.output_spec.shape[0],
+                output_size=1), alf.layers.Reshape(shape=()))
+
+        # +------------------------------------------+
+        # | Step 4: Assemble network + value head    |
+        # +------------------------------------------+
+
+        encoder_output_size = self._actor_encoder.output_spec.shape[0]
+
+        if is_sharing_encoder:
+            self._composition = alf.nn.Sequential(
+                self._actor_encoder,
+                alf.nn.Branch(
+                    self._policy_head,
+                    alf.nn.Sequential(
+                        # Use the same encoder, but the encoder is DETACHED.
+                        alf.layers.Detach(),
+                        alf.layers.FC(
+                            input_size=encoder_output_size, output_size=1),
+                        alf.layers.Reshape(shape=()),
+                        input_tensor_spec=self._actor_encoder.output_spec),
+                    self._aux_head))
+        else:
+            # When not sharing encoder, create a separate encoder for the value
+            # component.
+            self._value_encoder = encoding_network_ctor(
+                input_tensor_spec=observation_spec,
+                kernel_initializer=kernel_initializer)
+
+            self._composition = alf.nn.Sequential(
+                alf.nn.Branch(
+                    alf.nn.Sequential(
+                        self._actor_encoder,
+                        alf.nn.Branch(
+                            self._policy_head,
+                            self._aux_head,
+                            name='PolicyComponent')),
+                    alf.nn.Sequential(
+                        self._value_encoder,
+                        alf.layers.FC(
+                            input_size=encoder_output_size, output_size=1),
+                        alf.layers.Reshape(shape=()))),
+                # Order: policy, value, aux value
+                lambda heads: (heads[0][0], heads[1], heads[0][1]))
+
+    def forward(self, observation, state):
+        """Computes the action distribution, aux value and value estimation
+
+        Args:
+
+            observation (nested torch.Tensor): a tensor that is consistent with
+                the encoding network
+
+            state: the state(s) for RNN based network
+
+        Returns:
+
+            output (Triplet): network output in the order of policy (action
+                distribution), value function estimation, auxiliary value
+                function estimation
+
+            state (Triplet): RNN states in the order of policy, value, aux value
+
+        """
+        return self._composition(observation, state=state)
```

Comment:
If you always output the three values, will there be redundant computation? For example, sometimes you might only need one or two of them.

Reply:
Valid point. Yep, as mentioned above this is one of the concern. However since currently the layer is rather thin, it should not account for noticeable overhead. 

On the other hand making up for this will make the logic a bit more complicated. Plan to add a `TODO` here - if this becomes a problem in the future, we know how to fix it.